File: manager/src/main/java/roro/stellar/manager/StellarSettings.java
package roro.stellar.manager;

import static java.lang.annotation.RetentionPolicy.SOURCE;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;

import androidx.annotation.IntDef;
import androidx.annotation.NonNull;

import java.lang.annotation.Retention;

import roro.stellar.manager.utils.EmptySharedPreferencesImpl;

public class StellarSettings {

    public static final String NAME = "settings";
    
    public static final String KEEP_START_ON_BOOT = "start_on_boot";
    
    public static final String KEEP_START_ON_BOOT_WIRELESS = "start_on_boot_wireless";
    
    public static final String TCPIP_PORT = "tcpip_port";
    
    public static final String TCPIP_PORT_ENABLED = "tcpip_port_enabled";
    
    public static final String THEME_MODE = "theme_mode";

    private static SharedPreferences sPreferences;

    public static SharedPreferences getPreferences() {
        return sPreferences;
    }

    @NonNull
    private static Context getSettingsStorageContext(@NonNull Context context) {
        Context storageContext = context.createDeviceProtectedStorageContext();
        storageContext = new ContextWrapper(storageContext) {
            @Override
            public SharedPreferences getSharedPreferences(String name, int mode) {
                try {
                    return super.getSharedPreferences(name, mode);
                } catch (IllegalStateException e) {
                    return new EmptySharedPreferencesImpl();
                }
            }
        };
        return storageContext;
    }

    public static void initialize(Context context) {
        if (sPreferences == null) {
            sPreferences = getSettingsStorageContext(context).getSharedPreferences(NAME, Context.MODE_PRIVATE);
            
            if (!sPreferences.contains(TCPIP_PORT_ENABLED)) {
                sPreferences.edit().putBoolean(TCPIP_PORT_ENABLED, true).apply();
            }
            if (!sPreferences.contains(TCPIP_PORT)) {
                int randomPort = 1000 + (int) (Math.random() * 9000);
                sPreferences.edit().putString(TCPIP_PORT, String.valueOf(randomPort)).apply();
            }
        }
    }

    @IntDef({LaunchMethod.UNKNOWN, LaunchMethod.ROOT, LaunchMethod.ADB,})
    @Retention(SOURCE)
    public @interface LaunchMethod {
        int UNKNOWN = -1;
        int ROOT = 0;
        int ADB = 1;
    }

    @LaunchMethod
    public static int getLastLaunchMode() {
        return getPreferences().getInt("mode", LaunchMethod.UNKNOWN);
    }

    public static void setLastLaunchMode(@LaunchMethod int method) {
        getPreferences().edit().putInt("mode", method).apply();
    }
}

File: manager/src/main/kotlin/roro/stellar/manager/StellarSettings.kt
package roro.stellar.manager

import android.content.Context
import android.content.ContextWrapper
import android.content.SharedPreferences
import roro.stellar.manager.util.EmptySharedPreferencesImpl
import roro.stellar.manager.util.PortBlacklistUtils

object StellarSettings {
    const val NAME = "settings"
    const val BOOT_SCRIPT_ENABLED = "boot_script_enabled"
    const val BOOT_MODE = "boot_mode"
    const val TCPIP_PORT = "tcpip_port"
    const val TCPIP_PORT_ENABLED = "tcpip_port_enabled"
    const val THEME_MODE = "theme_mode"
    const val START_PAGE = "start_page"
    const val DROP_PRIVILEGES = "drop_privileges"
    const val SHIZUKU_COMPAT_ENABLED = "shizuku_compat_enabled"
    const val ACCESSIBILITY_AUTO_START_PROMPTED = "accessibility_auto_start_prompted"
    const val LAST_VERSION_CODE = "last_version_code"
    const val DAEMON_ENABLED = "daemon_enabled"

    /** 开机启动模式，三选一，默认 NONE */
    enum class BootMode { NONE, BROADCAST, ACCESSIBILITY, SCRIPT }

    /** 记录上次服务的启动方式（Root / ADB），用于开机自启时决定走哪条路径 */
    enum class LaunchMethod { UNKNOWN, ROOT, ADB }
    const val LAST_LAUNCH_METHOD = "last_launch_method"

    fun getLastLaunchMethod(): LaunchMethod {
        val name = getPreferences().getString(LAST_LAUNCH_METHOD, LaunchMethod.UNKNOWN.name)
            ?: LaunchMethod.UNKNOWN.name
        return runCatching { LaunchMethod.valueOf(name) }.getOrDefault(LaunchMethod.UNKNOWN)
    }

    fun setLastLaunchMethod(method: LaunchMethod) {
        getPreferences().edit().putString(LAST_LAUNCH_METHOD, method.name).apply()
    }

    fun getBootMode(): BootMode {
        val name = getPreferences().getString(BOOT_MODE, BootMode.NONE.name) ?: BootMode.NONE.name
        return runCatching { BootMode.valueOf(name) }.getOrDefault(BootMode.NONE)
    }

    fun setBootMode(mode: BootMode) {
        getPreferences().edit().putString(BOOT_MODE, mode.name).apply()
    }

    private var preferences: SharedPreferences? = null

    fun getPreferences(): SharedPreferences = preferences ?: EmptySharedPreferencesImpl()

    private fun getSettingsStorageContext(context: Context): Context {
        val storageContext = context.createDeviceProtectedStorageContext()
        return object : ContextWrapper(storageContext) {
            override fun getSharedPreferences(name: String?, mode: Int): SharedPreferences {
                return try {
                    super.getSharedPreferences(name, mode)
                } catch (_: IllegalStateException) {
                    EmptySharedPreferencesImpl()
                }
            }
        }
    }

    fun initialize(context: Context) {
        if (preferences == null) {
            preferences = getSettingsStorageContext(context)
                .getSharedPreferences(NAME, Context.MODE_PRIVATE)

            preferences?.let { prefs ->
                if (!prefs.contains(TCPIP_PORT_ENABLED)) {
                    prefs.edit().putBoolean(TCPIP_PORT_ENABLED, true).apply()
                }
                if (!prefs.contains(TCPIP_PORT)) {
                    var randomPort = PortBlacklistUtils.generateSafeRandomPort(1000, 9999, 100)
                    if (randomPort == -1) {
                        randomPort = 8765
                    }
                    prefs.edit().putString(TCPIP_PORT, randomPort.toString()).apply()
                }
            }
        }
    }
}
